docs(plugins): more in-depth pluginOptionsSchema documentation

mxstbr opened the pull request:
This expands upon the initial docs added in #27337 with a dedicated page for "Validating plugin options".

[ch16681]

mxstbr commented on docs/docs/configuring-usage-with-plugin-options.md:
```diff
@@ -54,24 +54,6 @@ The following table lists possible options values and an example plugin that mak
 
 **Note**: Themes (which are a type of plugin) are able to receive options from a site's `gatsby-config` to be used in its `gatsby-config` in order to allow themes to be composed together. This is done by exporting the `gatsby-config` as a function instead of an object. You can see an example of this in the [`gatsby-theme-blog`](https://github.com/gatsbyjs/themes/tree/master/packages/gatsby-theme- ...
 
-## How to validate options
```
This entire section was moved to its own page, and (on the website) is liked right at the end of this page since it is the next page defined in `doc-links.yml`

mxstbr commented on docs/docs/validating-plugin-options.md:
```diff
@@ -0,0 +1,230 @@
+
+To define their options schema, plugins export a [`pluginOptionsSchema`](/docs/node-apis/#pluginOptionsSchema) from their `gatsby-node.js`. This function gets passed an instance of [Joi](https://joi.dev) to define the schema with.
+
+For example, consider the following plugin called `gatsby-plugin-console-log`:
```
This is the same example used throughout the "Configuring usage with plugin options" doc, which is right before this one, however I figured it wouldn't hurt to re-introduce it since folks might open this page directly.

meganesu replied:
Agreed! Good thinking 👏

mxstbr commented on docs/docs/validating-plugin-options.md:
```diff
@@ -0,0 +1,230 @@
+  }).external(async pluginOptions => {
+    try {
+      await contentful
+        .createClient({
```
Keen eyes will spot that this is not the actual code used in the Contentful plugin. That is on purpose, as the actual code uses a vanilla `fetch` call with headers etc. and is more verbose, i.e. makes for a bad example as it hides the true intent of the function.

I figured the contentful SDK API would make the example clearer!

meganesu replied:
It might be helpful to put that caveat in here, so that people don't use this as a source of truth re: how Contentful works. It might be as simple as just changing the paragraph introducing the example to something like, "For example, this is how the Contentful source plugin schema might look if they wanted to validate using [whatever thing they would use instead of `fetch`]"

LekoArts commented on docs/docs/validating-plugin-options.md:
````diff
@@ -0,0 +1,230 @@
+
+1. Add the `gatsby-plugin-utils` package to your site:
+
+   ```shell
+   yarn add gatsby-plugin-utils
+   ```
+
+   or
+
+   ```shell
+   npm install gatsby-plugin-utils
+   ```
````
````suggestion
   ```shell
   npm install gatsby-plugin-utils
   ```
````

We're only ever using `npm` in our docs :)

LekoArts commented on docs/docs/validating-plugin-options.md:
````diff
@@ -0,0 +1,230 @@
+
+2. Use the `testPluginOptionsSchema` function exported from it in your test file. For example, with [Jest](https://jestjs.io):
+
+   ```javascript:title=plugins/gatsby-plugin-console/__tests__/gatsby-node.test.js
````
```suggestion
   ```javascript:title=plugins/gatsby-plugin-console/__tests__/gatsby-node.js
```

Our own guide leaves out the `.test` part when using `__tests__` folder, so I guess keep it consistent?
https://www.gatsbyjs.com/docs/unit-testing/

LekoArts commented on docs/docs/validating-plugin-options.md:
```diff
@@ -0,0 +1,230 @@
+
+### Add descriptions
+
+Make sure that every option and field has a `.description()` explaining its purpose. While this is not immediately visible, there might be tooling in the future that auto-generates plugin option documentation from the schema.
```
Maybe add that this also helps with self-documenting what the option is for the developers themselves?

LekoArts commented on docs/docs/validating-plugin-options.md:
```diff
@@ -0,0 +1,230 @@
+
+### Add custom error messages
+
+Sometimes you might want to provide more detailed error messages when validation fails for a specific field. Joi provides a [`.messages()` method](https://github.com/sideway/joi/issues/2109) which lets you override error messages for specific [error types](https://joi.dev/api/?v=17.2.1#list-of-errors).
```
Why not link to https://joi.dev/api/?v=17.2.1#anymessagesmessages in first link?

mxstbr replied:
Because the documentation for that function is practically useless. 😬  It shows this:

<img width="530" alt="Screenshot 2020-10-20 at 08 33 24" src="https://user-images.githubusercontent.com/7525670/96548903-fb9e5800-12ae-11eb-8357-9ab9e7c893a5.png">

which links to

<img width="563" alt="Screenshot 2020-10-20 at 08 33 37" src="https://user-images.githubusercontent.com/7525670/96548917-00fba280-12af-11eb-9cd9-e8e61762a452.png">

neither of which tell the user anything about using `.messages()`. I'd rather link something that has a useful explanation that helps plugin authors dive deeper!

LekoArts replied:
But if the API gets changed the link to that issue is obsolete and we won't know it until someone runs into problems. You can add a note that it takes in the same object like `messages` on the `validate` function: https://joi.dev/api/?v=17.2.1#anyvalidatevalue-options

LekoArts commented on www/src/data/sidebars/doc-links.yaml:
```diff
@@ -297,6 +297,8 @@
                   link: /docs/creating-a-transformer-plugin/
                 - title: Configuring Usage with Plugin Options
                   link: /docs/configuring-usage-with-plugin-options/
+                - title: Validating plugin options
```
You'll need to update this in the mansion repo

mxstbr replied:
Wait, for real? Then we should delete this file from this repo. :facepalm:

LekoArts replied:
It's something that @jxnblk is aware of

LekoArts reviewed (changes requested):
Left some comments :)

laurieontech commented on docs/docs/validating-plugin-options.md:
```diff
@@ -0,0 +1,230 @@
+
+## How to define an options schema
+
+To define their options schema, plugins export a [`pluginOptionsSchema`](/docs/node-apis/#pluginOptionsSchema) from their `gatsby-node.js`. This function gets passed an instance of [Joi](https://joi.dev) to define the schema with.
```
```suggestion
To define their options schema, plugins export [`pluginOptionsSchema`](/docs/node-apis/#pluginOptionsSchema) from their `gatsby-node.js` file. This function gets passed an instance of [Joi](https://joi.dev) to define the schema with.
```

laurieontech commented on docs/docs/validating-plugin-options.md:
````diff
@@ -0,0 +1,230 @@
+   npm install gatsby-plugin-utils
+   ```
+
+2. Use the `testPluginOptionsSchema` function exported from it in your test file. For example, with [Jest](https://jestjs.io):
````
What is "it" here?

laurieontech commented on docs/docs/validating-plugin-options.md:
```diff
@@ -0,0 +1,230 @@
+
+### Set default options
+
+You can use the `.default()` method to default a value for an option. For example with `gatsby-plugin-console-log`, this is how it can default the `message` option to `"default message"` in case users do not pass their own `message`:
```
```suggestion
You can use the `.default()` method to set a default value for an option. For example with `gatsby-plugin-console-log`, you can have the `message` option default to  `"default message"` in case a user does not pass their own `message`:
```

laurieontech commented on docs/docs/validating-plugin-options.md:
```diff
@@ -0,0 +1,230 @@
+
+### Validating external access
+
+Some plugins (particularly source plugins) query external APIs. With the `.external()` method, they can asynchronously validate that users have access to the API to provide a better experience if they pass invalid secrets.
```
```suggestion
Some plugins (particularly source plugins) query external APIs. With the `.external()` method, you can asynchronously validate that the user has access to the API, providing a better experience if they pass invalid secrets.
```

laurieontech reviewed (commented):
Left a handful of comments. It's hard to explain the usefulness of this with just text. Instead, I might start this doc off with a screenshot example of an invalid option and then backtrack to explain how that came to exist.

MichaelDeBoey commented on docs/docs/validating-plugin-options.md:
````diff
@@ -0,0 +1,224 @@
+1. Add the `gatsby-plugin-utils` package to your site:
+
+   ```shell
+   npm install gatsby-plugin-utils
````
This should be a `devDependency` I think? 🤔

```suggestion
   npm install --dev gatsby-plugin-utils
```

MichaelDeBoey commented on docs/docs/validating-plugin-options.md:
```diff
@@ -0,0 +1,224 @@
+
+Make sure that every option and field has a `.description()` explaining its purpose. This is helpful for documentation as users can look at the schema and understand all the options. There might also be tooling in the future that auto-generates plugin option documentation from the schema.
+
+### Set default options
```
This makes it look like it's preferred to set default values on the schema over using JavaScript default values.

```js
exports.onPreInit = (_, { message = `default message` }) => {
  console.log(`logging: "${message}" to the console`)
}
```

Any reason for this preference? 🤔

mxstbr replied:
Yes. If you set default values using JavaScript, they will only apply to one lifecycle:

```JS
exports.onPreInit = (_, { message = "default" }) => {
  ...
}

exports.sourceNodes = (_, { message }) => {
  // Here message won't be defaulted.
}
```

MichaelDeBoey commented on docs/docs/validating-plugin-options.md:
```diff
@@ -0,0 +1,224 @@
+          accessToken: pluginOptions.accessToken,
+        })
+        .getSpace()
+    } catch (err) {
```
```suggestion
    } catch {
```

meganesu commented on docs/docs/validating-plugin-options.md:
```diff
@@ -0,0 +1,224 @@
+
+Sometimes you might want to provide more detailed error messages when validation fails for a specific field. Joi provides a [`.messages()` method](https://joi.dev/api/?v=17.2.1#anymessagesmessages) which lets you override error messages for specific [error types](https://joi.dev/api/?v=17.2.1#list-of-errors) (e.g. `"any.required"` when a `.required()` call fails).
+
+For example, this is how we would provide a custom error message if users do not specify `optionA` for `gatsby-plugin-console-log`:
```
Per the style guide, [use "you" instead of "we"](https://www.gatsbyjs.com/contributing/gatsby-style-guide/#use-you-as-the-pronoun).

```suggestion
For example, in the `gatsby-plugin-console-log` plugin above, this is how you would provide a custom error message if users do not specify `optionA`:
```

meganesu commented on docs/docs/validating-plugin-options.md:
```diff
@@ -0,0 +1,224 @@
+
+### Deprecating options
+
+Rather than removing options from the schema, deprecate them using the [`.forbidden()` method](https://joi.dev/api/?v=17.2.1#anyforbidden) in a major version release. Then, [add a custom error message](#add-custom-error-messages) explaining how users should upgrade the functionality using `.messages()`.
```
Why? What do they lose by just removing options?

meganesu commented on docs/docs/validating-plugin-options.md:
```diff
@@ -0,0 +1,224 @@
+
+## How to define an options schema
+
+To define their options schema, plugins export [`pluginOptionsSchema`](/docs/node-apis/#pluginOptionsSchema) from their `gatsby-node.js` file. This function gets passed an instance of [Joi](https://joi.dev) to define the schema with.
```
Could this get converted into a numbered list of steps? This feels like the core idea behind this "how to" guide, and right now it's a little sparse. Someone might skip over it without realizing that those are the steps they're meant to be following.

mxstbr replied:
I tried that but it felt very strange, as it's really only one step 😅 At least I made this paragraph a tiny bit meatier...

meganesu commented on docs/docs/configuring-usage-with-plugin-options.md:
```diff
@@ -54,24 +54,6 @@ The following table lists possible options values and an example plugin that mak
 
```
Upon rereading this doc, it might make more sense for the "what can be passed in as options" section to come before the "where to access plugin options" section. So that it flows from why you may want plugin options → what kinds of options you can use → how to set up those options.

meganesu commented on docs/docs/validating-plugin-options.md:
````diff
@@ -0,0 +1,224 @@
+   npm install gatsby-plugin-utils
+   ```
+
+2. Use the `testPluginOptionsSchema` function exported from the package in your test file. For example, with [Jest](https://jestjs.io):
````
```suggestion
2. Use the `testPluginOptionsSchema` function exported from the package in your test file. For example, with [Jest](https://jestjs.io), your tests might look something like this:
```

meganesu replied:
Could we link to something that shows the API for `testPluginOptionsSchema`? Or at least briefly explain what the parameters are?

e.g., "This function takes two parameters:
* the plugin's actual Joi schema
* an example plugin options object to test

It returns an object with an `isValid` key, which will be `true` or `false` based on whether or not the options object fits the actual Joi schema."

(Actually, maybe it makes more sense for that API to live in the gatsby-plugin-utils readme?)

meganesu commented on docs/docs/validating-plugin-options.md:
````diff
@@ -0,0 +1,224 @@
+Accessing `pluginOptions.message` would then log `"default message"` in all plugin APIs if the user does not supply their own value:
+
+```javascript:title=plugins/gatsby-plugin-console-log/gatsby-node.js
+exports.onPreInit = (_, pluginOptions) => {
````
Sidebar question: Are plugin methods only available from onPreInit, or is it in any of those APIs?

> `exports.onPreInit = (_, pluginOptions)`

Also what is the `_` parameter syntax? I haven't seen that before and couldn't figure out how to google it.

mxstbr replied:
The pluginOptions are available in all of the APIs!

> Also what is the _ parameter syntax? I haven't seen that before and couldn't figure out how to google it.

This is a convention I thought was common to denote unused arguments. (`(_, pluginOptions)` === the first argument to the function isn't used, `(input, pluginOptions)` === it is used) However, that doesn't make any sense in documentation, I should just put the actual name there. Good call! 👍

meganesu reviewed (changes requested):
Thanks for taking the time to write this up! I love all the examples, I think those will be super helpful for readers.

For transparency, I'm thinking about moving toward structuring our docs more like this [Divio documentation system](https://documentation.divio.com/). This particular doc feels like it would fall under the [How-To Guide](https://documentation.divio.com/how-to-guides/) category, so a lot of my feedback is related to that.

Let me know if you have any questions!

mxstbr commented on docs/docs/validating-plugin-options.md:
```diff
@@ -0,0 +1,224 @@
+
+### Add descriptions
+
+Make sure that every option and field has a `.description()` explaining its purpose. This is helpful for documentation as users can look at the schema and understand all the options. There might also be tooling in the future that auto-generates plugin option documentation from the schema.
```
Tooling that we might create, and definitely a nice-to-have for the future without any concrete plans!

mxstbr commented on docs/docs/validating-plugin-options.md:
```diff
@@ -0,0 +1,224 @@
+
+Here are some specific Joi best practices for `pluginOptionsSchema`s:
+
+- [Add descriptions](#add-descriptions)
```
I was thinking of the list is a concise way to summarize the best practices for people to refer (back) to, rather than being a set of paragraph headers. However, I'm fine with removing it if you prefer that!

meganesu replied:
Makes sense to me, let's leave it!

mxstbr commented:
@meganesu thank you for the fantastic review! I have updated the doc based on your feedback, replied inline to clarify a few comments (the ones that I haven't marked as "Resolved" above) and moved the doc back into "Configuring usage with plugin options".

This is ready for another round of reviews! :pray:

MichaelDeBoey commented on docs/docs/configuring-usage-with-plugin-options.md:
```diff
@@ -54,24 +54,220 @@ The following table lists possible options values and an example plugin that mak
+
+#### Add custom error messages
+
+Sometimes you might want to provide more detailed error messages when validation fails for a specific field. Joi provides a [`.messages()` method](https://joi.dev/api/?v=17.2.1#anymessagesmessages) which lets you override error messages for specific [error types](https://joi.dev/api/?v=17.2.1#list-of-errors) (e.g. `"any.required"` when a `.required()` call fails).
```
```suggestion
Sometimes you might want to provide more detailed error messages when validation fails for a specific field. Joi provides a [`.messages()` method](https://joi.dev/api/?v=17.3.0#anymessagesmessages) which lets you override error messages for specific [error types](https://joi.dev/api/?v=17.3.0#list-of-errors) (e.g. `"any.required"` when a `.required()` call fails).
```

MichaelDeBoey commented on docs/docs/configuring-usage-with-plugin-options.md:
```diff
@@ -54,24 +54,220 @@ The following table lists possible options values and an example plugin that mak
+
+#### Deprecating options
+
+While you can simply remove options from the schema in major versions, that causes cryptic error messages for users upgrading with existing configuration. Instead, deprecate them using the [`.forbidden()` method](https://joi.dev/api/?v=17.2.1#anyforbidden) in a major version release. Then, [add a custom error message](#add-custom-error-messages) explaining how users should upgrade the functionali ...
```
```suggestion
While you can simply remove options from the schema in major versions, that causes cryptic error messages for users upgrading with existing configuration. Instead, deprecate them using the [`.forbidden()` method](https://joi.dev/api/?v=17.3.0#anyforbidden) in a major version release. Then, [add a custom error message](#add-custom-error-messages) explaining how users should upgrade the functionality using `.messages()`.
```

MichaelDeBoey commented on docs/docs/configuring-usage-with-plugin-options.md:
````diff
@@ -54,24 +54,220 @@ The following table lists possible options values and an example plugin that mak
+1. Add the `gatsby-plugin-utils` package to your site:
+
+   ```shell
+   npm install gatsby-plugin-utils
````
This should be a `devDependency` I think? 🤔

```suggestion
   npm install --dev gatsby-plugin-utils
```

MichaelDeBoey commented on docs/docs/configuring-usage-with-plugin-options.md:
```diff
@@ -54,24 +54,220 @@ The following table lists possible options values and an example plugin that mak
+          accessToken: pluginOptions.accessToken,
+        })
+        .getSpace()
+    } catch (err) {
```
```suggestion
    } catch {
```

meganesu commented on docs/docs/configuring-usage-with-plugin-options.md:
```diff
@@ -54,24 +54,229 @@ The following table lists possible options values and an example plugin that mak
+
+The [Joi API documentation](https://joi.dev/api/) is a great reference to use while working on a `pluginOptionsSchema`, as it shows all the available types and methods.
+
+Here are some specific Joi best practices for `pluginOptionsSchema`s:
```
Having the trailing "s" looked funny to me

```suggestion
Here are some specific Joi best practices for `pluginOptionsSchema`:
```

meganesu reviewed (approved):
This is beautiful! ✨ Great work! 👏

mxstbr commented:
> This is beautiful! ✨ Great work! 👏

Thank you, but that is just as much your doing as mine! Wouldn't have been possible without your thoughtful feedback. :pray:
